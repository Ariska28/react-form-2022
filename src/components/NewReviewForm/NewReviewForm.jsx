import styles from "./styles.module.css";
import classnames from "classnames";
import { useReducer } from "react";
import { Rating } from "../Rating/Rating";


const INITIAL_FORM_VALUE = {
  name: {value: "", touched: false, hasError: false },
  text: {value: "", touched: false, hasError: false },
  email: {value: "", touched: false, hasError: false },
  rating: 1,
};

const ERROR_MESSAGES = {
  empty: "Requared field",
  maxLength: "Uncorrect string length: max length 5",
  email: "Input correct email"
}

const ACTIONS = {
  setName: "setName",
  setText: "setText",
  setEmail: "setEmail",
  setRating: "setRating",
  clear: "clear",
};

const formReducer = (state, action) => {
  switch(action.type) {
    case ACTIONS.setName:  {
      return {
        ...state, 
        name: { 
          value: action.payload, 
          touched: true,
          ...validation.isEmpty(action.payload), 
          ...validation.maxLength(action.payload) 
        }
      }
    }

    case ACTIONS.setText: {
      return {
        ...state, 
        text: {
          value: action.payload,
          touched: true,
          ...validation.isEmpty(action.payload),
        } 
      }
    }

    case ACTIONS.setEmail: {
      return {
        ...state, 
        email: {
          value: action.payload,
          touched: true,
          ...validation.isEmpty(action.payload),
          ...validation.isCorrectEmail(action.payload),
        } 
      }
    }

    case ACTIONS.setRating: {
      return {
        ...state,
        rating: action.payload
      }
    }

    case ACTIONS.clear: {
      return INITIAL_FORM_VALUE;
    }

    default: 
      return state;
  }
};

const validation = {
  isEmpty: (value) => {
    if (value.length < 1) {
      console.log(new Error(ERROR_MESSAGES.empty));
      return { hasError: true }
    }
  },
  maxLength: (value) => {
    if (value.length > 5) {
      console.log(new Error(ERROR_MESSAGES.maxLength));
      return { hasError: true }
    }  
  },
  isCorrectEmail: (value) => {
    console.log(new Error(ERROR_MESSAGES.email));
    if(!(/^\w+([\.-]?\w+)*@\w+([\.-]?\w+)*(\.\w{2,3})+$/.test(value))) {
      return { hasError: true }
    }
  }
}

const addValidationStyles = (hasError, touched) => {
  if (hasError) {
    return styles.error;
  }

  if(touched) {
    return styles.succes;
  }
}

export const NewReviewForm = ({ className }) => {
  const [formValue, dispatch] = useReducer(formReducer, INITIAL_FORM_VALUE);

  const onNameChange = (event) => {
    dispatch({ type: ACTIONS.setName, payload: event.target.value });
  };
  const onTextChange = (event) => {
    dispatch({ type: ACTIONS.setText, payload: event.target.value });
  };
  const onEmailChange = (event) => {
    dispatch({ type: ACTIONS.setEmail, payload: event.target.value });
  };
  const onRatingChange = (value) => {
    dispatch({ type: ACTIONS.setRating, payload: value });
  };

  return (
    <div className={classnames(styles.root, className)}>
      <div className={classnames(styles.formControl, addValidationStyles(formValue.name.hasError, formValue.name.touched))}>
        <label htmlFor="name">Name</label>
        <input
          name="name"
          type="text"
          value={formValue.name.value}
          onChange={onNameChange}
        />
      </div>
      <div className={classnames(styles.formControl, addValidationStyles(formValue.text.hasError, formValue.text.touched))}>
        <label htmlFor="text">Text</label>
        <input
          name="text"
          type="text"
          className={styles.input}
          value={formValue.text.value}
          onChange={onTextChange}
        />
      </div>
      <div className={classnames(styles.formControl, addValidationStyles(formValue.email.hasError, formValue.email.touched))}>
        <label htmlFor="text">Email</label>
        <input
          name="text"
          type="email"
          className={styles.input}
          value={formValue.email.value}
          onChange={onEmailChange}
        />
      </div> 
      <div className={styles.formControl}>
        <label>Choose rating</label>
        <Rating value={formValue.rating} onChange={onRatingChange} />
      </div>
      <button className={styles.clear} onClick={() => dispatch({ type: ACTIONS.clear })}>Clear</button>
    </div>
  );
};
