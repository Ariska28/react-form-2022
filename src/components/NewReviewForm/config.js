export const STATE = {
  DEFAULT: 'DEFAULT',
  ERROR: 'ERROR',
  SUCCESS: 'SUCCESS',
  DISABLED: 'DISABLED',
  READONLY: 'READONLY'
}

export const TYPE = {
  TEXT: "text",
  NAME: "name",
  EMAIL: "email"
}

export const MESSAGE = {
  ERROR: "The field is invalid",
  SUCCESS: "Great!"
}

export const STATUS_MESSAGE_MAP = {
  [STATE.DEFAULT] : "",
  [STATE.ERROR] : MESSAGE.ERROR,
  [STATE.SUCCESS] : MESSAGE.SUCCESS,
  [STATE.DISABLED] : "",
  [STATE.READONLY]: ""
}

export const INITIAL_FORM_VALUE = {
  name: "",
  text: "",
  email: "",
};

export const FORM_MESSAGES = {
  default: "",
  error: <span className="error">Form has error, please check it</span>,
  success: <span className="success">Everething is cool</span>,
  empty: <span className="error">Some fileds don't filled, please fix it</span>
}
