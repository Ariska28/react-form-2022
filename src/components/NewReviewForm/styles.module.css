.root {
    display: flex;
    flex-direction: column;
}

.formControl {
    display: flex;
    flex-wrap: wrap;
    align-items: center;
}

.formControl input {
    height: 30px;
    border: 1px solid var(--color-primary);
    border-radius: 5px;
    width: 100%;
    padding-left: 10px;
    outline: none;
}

.error input {
    border: 2px solid var(--color-error);
}

.succes input {
    border: 2px solid var(--color-success);
}

.formControl label {
    margin-bottom: 8px;
    font-size: 18px;
    color: var(--color-header-bg);
}

.formControl:not(:last-child) {
    margin-bottom: 20px;
}

.clear {
    max-width: 200px;
    padding: 10px 20px;
    border-radius: 5px;
    border: 1px solid var(--color-text);
    color: var(--color-text);
    background-color: var(--color-white);
    font-weight: 600;
}

.clear:hover {
    color: var(--color-white);
    background-color: var(--color-text);
}

.clear:focus {
    color: var(--color-white);
    background-color: var(--color-text);
}