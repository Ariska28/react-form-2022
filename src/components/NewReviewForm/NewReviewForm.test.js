import React from "react";
import renderer from "react-test-renderer";
import {NewReviewForm, FormElementPicker, FormElement} from './NewReviewForm';
import {BrowserRouter as Router} from 'react-router-dom';

describe("Jest Snapshot test newReviewForm component", () => {
  it("Match to DOM Snapshot", () => {
    const domTree = renderer.create(<Router><NewReviewForm /></Router>).toJSON();
    expect(domTree).toMatchSnapshot();
  })
});

describe("Jest Snapshot test FormElementPicker component", () => {
  it("Match to DOM Snapshot", () => {
    const domTree = renderer.create(<FormElementPicker />).toJSON();
    expect(domTree).toMatchSnapshot();
  })
})

describe("Jest Snapshot test FormElement component", () => {
  it("Match to DOM Snapshot", () => {
    const domTree = renderer.create(<FormElement />).toJSON();
    expect(domTree).toMatchSnapshot();
  })
})



