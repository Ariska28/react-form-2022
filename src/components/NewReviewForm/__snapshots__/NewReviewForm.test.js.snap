// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Jest Snapshot test FormElement component Match to DOM Snapshot 1`] = `
<div
  className="formControl"
>
  <label />
  <input
    type="text"
  />
</div>
`;

exports[`Jest Snapshot test FormElementPicker component Match to DOM Snapshot 1`] = `
<div
  className="formControl"
>
  <label>
    Choose 
  </label>
  <div
    className="container"
  >
    <div>
      <img
        className="star m"
        loading="lazy"
        onClick={[Function]}
        src="star-gold.svg"
      />
      <img
        className="star m"
        loading="lazy"
        onClick={[Function]}
        src="star-gold.svg"
      />
      <img
        className="star m"
        loading="lazy"
        onClick={[Function]}
        src="star-gold.svg"
      />
      <img
        className="star m"
        loading="lazy"
        onClick={[Function]}
        src="star-gold.svg"
      />
      <img
        className="star m"
        loading="lazy"
        onClick={[Function]}
        src="star-gold.svg"
      />
    </div>
  </div>
  <input
    className="hidden"
  />
</div>
`;

exports[`Jest Snapshot test newReviewForm component Match to DOM Snapshot 1`] = `
<form
  className="root"
  onSubmit={[Function]}
>
  <div
    className="formControl"
  >
    <label
      htmlFor="name"
    >
      name
    </label>
    <input
      name="name"
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
  <div
    className="formControl"
  >
    <label
      htmlFor="text"
    >
      text
    </label>
    <input
      name="text"
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
  <div
    className="formControl"
  >
    <label
      htmlFor="email"
    >
      email
    </label>
    <input
      name="email"
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
  <div
    className="formControl"
  >
    <label>
      Choose 
      rating
    </label>
    <div
      className="container"
    >
      <div>
        <img
          className="star m"
          loading="lazy"
          onClick={[Function]}
          src="star-gold.svg"
        />
        <img
          className="star m"
          loading="lazy"
          onClick={[Function]}
          src="star.svg"
        />
        <img
          className="star m"
          loading="lazy"
          onClick={[Function]}
          src="star.svg"
        />
        <img
          className="star m"
          loading="lazy"
          onClick={[Function]}
          src="star.svg"
        />
        <img
          className="star m"
          loading="lazy"
          onClick={[Function]}
          src="star.svg"
        />
      </div>
    </div>
    <input
      className="hidden"
      name="rating"
      onChange={[Function]}
      type="number"
      value={1}
    />
  </div>
  <div
    className="buttons"
  >
    <button
      className="clear"
      onClick={[Function]}
      type="submit"
    >
      Send
    </button>
    <button
      className="clear"
      onClick={[Function]}
    >
      Clear
    </button>
  </div>
</form>
`;
